Embed and host images

olivier-vadiaval opened the pull request:
Fix #16 #41 #42
Partly resolves #36 but still need to display unlisted posts
Allow image hosting in media folder
Allow image embedding through markdown
Add an Image model
Allow unlisted post embedding

We can now embed __unlisted__ posts with images into other posts with markdown using their url (http://127.0.0.1:8000/author/{author_id}/posts/{post_id})

debecerra commented on code/socialDistribution/models/post.py:
```diff
@@ -85,6 +85,12 @@ class Visibility(models.TextChoices):
     #     blank=True,
     # )
 
+    image = models.ForeignKey('Image',
+        blank=True,
+        null=True,
+        on_delete=models.CASCADE
+    )
+
```
Can't we save the image as a base64 encoding in the content BinaryField? I only say this because InboxPosts (posts that are received by the user) don't have this field.

olivier-vadiaval replied:
I think Michelle told me about that at some point but to be honest, this completely escaped my mind. 😅 I'll make the changes

I'm also wondering about something related to that. Since we allow the user to upload image files on top of having content text, should we create a separate post for holding that image and then embed the url to that image post at the end of the content (text) of the first post. If we do that, we can get rid of references to images inside the Post models and store every content as binary.

When the post is rendered, the js code handling markdown will make a GET request for the uploaded image post as for regular markdown.

Does that make sense?

debecerra replied:
> I'm also wondering about something related to that. Since we allow the user to upload image files on top of having content text, should we create a separate post for holding that image and then embed the url to that image post at the end of the content (text) of the first post. If we do that, we can get rid of references to images inside the Post models and store every content as binary.

Yes, exactly. I think storing everything as binary is the way to go. OR ( because it's simpler), we just force the user to make either a text post OR an image post, but not both. That way the user has to make two posts on their own and manually embed whatever image they want to add. I think this would be quicker and easier.



> When the post is rendered, the js code handling markdown will make a GET request for the uploaded image post as for regular markdown.

Not sure if I totally understand, but yes I think so. The markdown will reference the image URL that is hosted by our server, or any other image URL that the user adds to their markdown post. I think it would be best if the image posts were hosted from a URL different from /author/id/post/id. The /post/id endpoint should probably only be for the JSON post objects with the base64 encoded content field.

e.g. you could make an /media/ path and serve the images as images from there (if you don't do that already, I'm just not totally sure if I understood.) GETing from this /media/{id} path could render the binary content of post with id={id} as an image, **only if that post has content_type == "image or whatever".**

At least those are my thoughts, you're free to see how to best get this working. Let me know if that doesn't make sense.

olivier-vadiaval replied:
Yup, you're right. I think I'll go with the second approach since it seems easier and cleaner (I think). That is, have allow the user to create only one type of post (either image or text/md).

debecerra commented on code/socialDistribution/views.py:
```diff
@@ -300,34 +300,65 @@ def author(request, author_id):
-            bin_content = form.cleaned_data.get('content_media')
-            if bin_content is not None:
-                content_media = base64.b64encode(bin_content.read())
+            media = form.cleaned_data.get('content_media')
+            content = form.cleaned_data.get('content_text')
+            if media is not None:
+                uploaded_img = Image.objects.create(
+                    caption=form.cleaned_data.get('description'),
+                    image=media
+                )
```
Can we store both content_media or content_text (which ever is filled in) in a `content` `BinaryField`? I talked to Michelle about this but I can send you what I sent her.

debecerra reviewed (changes requested):
I think it's best if we continue storing the media as a BinaryField. Otherwise, I think it becomes too difficult to handle differences between local posts and remote posts in the models. I had some ideas I ran by Michelle since I thought she was working on this, let me know if you want to discuss it.

Also, is this able to handle different visibility levels of posts?

Thanks for the hard work and for taking a shot at this!

mxiuwang reviewed (commented):
Do we not have to store images in base64? I agree binary is good but isn't base64 part of the spec

mxiuwang reviewed (changes requested):
Everything looks good! If we are not allowing users to upload images (content_media) in text or markdown posts, I think we should remove this field when making a post

olivier-vadiaval commented:
> Everything looks good! Users cannot upload images to "text or markdown" post type. I'm not sure if this is intentional. If we are not allowing users to upload images (content_media) in text or markdown posts, I think we should remove this field.

I think I removed every content_media field 🤔, except from a migration file (which we shouldn't directly modify). I've basically removed the content_media and content_text, and added a content field which stores the post content as base64-encoded binary data. Please let me know if I've missed any lingering content_media variable during the refactoring, thanks! 🙏

debecerra commented on code/socialDistribution/views.py:
```diff
@@ -308,34 +308,40 @@ def author(request, author_id):
 def create(request):
     return render(request, 'create/index.html')
 
+def post(request, author_id, post_id):
+    """
+        Allows user to view a specific post.
+        Mostly useful to share public (unlisted) posts through url
+    """
+    author = get_object_or_404('Author', pk=author_id)
+    post = get_object_or_404('Post', pk=post_id)
+    pass
+
```
What does this do? It throws an error since we don't return an HTTP object.

olivier-vadiaval replied:
Oh sorry, I was gonna address showing unlisted posts with this view function, but I'm gonna cut this pr here.
Michelle will be dealing with displaying the unlisted posts.
I'll remove this function anyways.

debecerra commented on code/socialDistribution/static/js/post_form.js:
```diff
@@ -0,0 +1,40 @@
+ */
+
+// form.post_type expands to an input tag with "id_post_type" as id
+var post_type = document.getElementById('id_post_type');
+var content_text_div = document.getElementById('content_text_div');
+var image_upload_div = document.getElementById('image_upload_div');
+function toggleContentInputHandler(ev) {
+    if (this.value == "TEXT") {
+        content_text_div.style.display = "block";
+        image_upload_div.style.display = "none";
+    }
+    else {
+        image_upload_div.style.display = "block";
+        content_text_div.style.display = "none";
+    }
+}
+
+toggleContentInputHandler.call(post_type);
+post_type.addEventListener('change', toggleContentInputHandler.bind(post_type));
+
+// form.visibility expands to an input tag with "id_visibility" as id
+var visibility_check = document.getElementById('id_visibility');
+var post_recipients_div = document.getElementById('post_recipients_div');
+function toggleRecipientsHandler (ev) {
+    if (this.value == "PR")
+        post_recipients_div.style.display = "block";
+    else
+        post_recipients_div.style.display = "none";
+}
+
+visibility_check.addEventListener('change', toggleRecipientsHandler.bind(visibility_check));
+toggleRecipientsHandler.call(visibility_check);
```
I don't think the toggle hiding on the form is working. I don't recall if we ever fixed the same issue for private posts either. 

After I create a single post (which works fine), the next time I open the form it doesn't hide as expected.

olivier-vadiaval replied:
I had this issue as well. For some reason, when I delete the db and re-migrate, it works. I'm still unsure of the root cause, but I think it's related to the modal from bootstrap interfering with this code.

debecerra commented on code/api/views.py:
```diff
@@ -231,6 +233,30 @@ def post(self, request, author_id):
+        post = get_object_or_404(LocalPost, pk=int(post_id))
+        accepted_types = request.headers['Accept']
+
+        if 'image' in accepted_types and post.is_image_post() and post.unlisted:
```
Is `'image' in accepted_types` necessary here? I feel like it might cause issues and be overly picky when responding to requests (thinking about other groups here).

Also, I had some issues finding the URL for the image I just created. Could you create some sort of print to console that says the ID of the unlisted post when created? We would ideally want some UI for this to tell the user where to find their unlisted post, but that can be left for another PR.

I also think it would be best to put this logic on a different route. What happens if I go to this URL in the browser. My intuition says I should get the post JSON object back since I just made a GET request. But then how do I see the rendered image if that's what I wanna do? Which will be given preference for a generic GET request at this URL (the JSON post object or the image)?

olivier-vadiaval replied:
I agree with making a new route. This endpoint wasn't implemented. What I had in mind was to re-use the post endpoint for returning the image if the "image/webp" mime-type was specified. But, I was thinking of giving priority to text/html and application/json later on. I made a new endpoint for a post's image.

Michelle told me she was working on a template for showing unlisted post. I asked her if she could add a button or some obvious way for the user to copy the route to the post's image to their clipboard. The built-in Clipboard api looks promising but I haven't had the chance to look further into it.

debecerra commented on code/cmput404/settings.py:
```diff
@@ -16,6 +16,8 @@
 # Build paths inside the project like this: BASE_DIR / 'subdir'.
 BASE_DIR = Path(__file__).resolve().parent.parent
 
+MEDIA_ROOT = os.path.join(BASE_DIR, 'media')
+MEDIA_URL = '/media/'
```
Are we still using this?

olivier-vadiaval replied:
Yup

debecerra commented on code/socialDistribution/forms.py:
```diff
@@ -83,4 +133,27 @@ def clean_visibility(self):
+
+        else:
+            content_type = LocalPost.ContentType.PLAIN
+            content = base64.b64encode(self.cleaned_data.get('content_text'))
```
Do we need to base64 encode text data? (just wondering)

olivier-vadiaval replied:
Hmm yes, you're right 🤔 I think we can just remove the b64encoding and just do an encode('utf-8') for storing since textual content won't be sent over as binary in any case. I don't know why I thought it would lol.  Thanks a lot for noticing that! ♥ This makes things easier.

debecerra commented on code/socialDistribution/forms.py:
```diff
@@ -83,4 +133,27 @@ def clean_visibility(self):
+        # Upload image to media folder and base64-
+        # encode the data to store in database
+        if post_type == PostForm.IMAGE and image:
+            Image.objects.create(
+                caption=self.cleaned_data.get('description'),
+                image=image
```
Why do we create an Image here? Do we still need the Image model?

olivier-vadiaval replied:
I'm using it to upload the images to a media folder like Jawad was suggesting during one of the previous meetings. I thought this was the agreed-upon functionality for #41 . 🤔

debecerra replied:
I thought we were storing the images as binary in the database (`content` field), not the media folder. I just don't see where the media folder is being used? But if there is a reason for it then that's fine too (as long as private posts are still private).

olivier-vadiaval replied:
I'll ask Jawad for further clarification. But, the only reason I see to keep it is it allows images not related to any post to be embedded in a post being created. Like with ![coolbears](http://127.0.0.1:8000/media/image/coolbears.jpg)

olivier-vadiaval replied:
Jawad clarified that part with me. You're right, we can the Image model. I made the changes in the latest commits and also moved the endpoint for GETting images from unlisted posts to the socialDistribution/views.py. See unlisted_post_image in socialDistribution/views.py. I'm still trying to figure out why the JS code isn't working properly but I'm unsure if I can get it fixed today.

olivier-vadiaval replied:
Thanks for the feedback! ♥

debecerra reviewed (changes requested):
Looking pretty good so far, but I did have some issues running it. 

Left a few comments about leftover code from the old implementation (I think) which should probably be deleted if not being used. Also left a few comments about the new features themselves. Some more comments throughout would be helpful as well.

Sorry if it's a lot I just have a lot of questions lol.

Thanks.

mxiuwang commented:
> content_media

I believe the issue is actually in the class PostForm - we need to enable to disable the "Upload File" button depending on the Post Type the user chooses. But dynamically updating a static form might require some redesign that's probably not a priority right now, so I can open an issue if we have time to solve it later?

debecerra commented on code/api/tests/test_posts.py:
```diff
@@ -18,7 +18,7 @@ def create_post(title, author):
             description="testDesc",
             content_type=LocalPost.ContentType.PLAIN,
-            content="testContexxt",
+            content=b64.b64encode("testContexxt".encode("utf-8")),
```
I fixed the failed test by removing the b64encoding. Why do we need to encode here? Won't data sent already be encoded if it's supposed to be encoded (like an image) and other data doesn't need to be encoded at all?

debecerra replied:
![image](https://user-images.githubusercontent.com/47841864/142777910-13035ca8-5a65-4e57-8c87-55032d68c2ae.png)
 
Image for reference. Data is getting encoded when it shouldn't be. So the post received by another author via their inbox is garbeled. Remove those b64encodes and I think this should be fixed.

debecerra commented on code/api/tests/test_inbox.py:
```diff
@@ -113,7 +119,7 @@ def test_post_local_post(self):
         self.assertEqual(inbox_post.title, dummy_post.title)
         self.assertEqual(inbox_post.description, dummy_post.description)
-        self.assertEqual(inbox_post.content, dummy_post.content)
+        self.assertEqual(inbox_post.decoded_content, dummy_post.decoded_content)
```
This line failed in tests. See other comment

olivier-vadiaval replied:
Sorry, I noticed the issue with the tests and was going to fix in a commit, together with the buggy js code for the form. But, I can't seem to find the root cause of the latter bug.

Thanks a lot for the comments!

Still working on the buggy js, but I'm going to focus on some of my other assigned tasks a bit more.

debecerra reviewed (approved):
Looks great, thanks for the hard work. 

Left one comment about a failing test and how I fixed it. Were you supposed to remove those instances of `base64.b64encode()`? I think there are three scattered around the code but I don't see why they are needed. 

Fix the failing test and go ahead and merge. Thanks

olivier-vadiaval commented:
> > content_media
> 
> I believe the issue is actually in the class PostForm - we need to enable to disable the "Upload File" field depending on the Post Type the user chooses. But dynamically updating a static form might require some redesign that's probably not a priority right now, so I can open an issue if we have time to solve it later?

I'll take a look at it soon. I haven't been able to pinpoint what's causing the toggle to stop working in the "Create post" form after editing a post. We might need some redesign as you said, but I think that I'll switch my focus to something else for now.
